### PR TITLE
[tailcall] Split up large tests for fine grained enable/disable, instead of large scale disable (F# and interface).

### PR DESCRIPTION
**Pull request by jaykrell**
Centralize tailcall debugprint.
Tweak tests to account for x86 frame alignment -- this way they pass, and in the real world, we might consume a little stack instead of zero, but not a lot. The test checks for zero.

### CODE REVIEW

**Comment by jaykrell** [comment]
Very relevant failures. So far not reproing locally but still trying.

**Comment by jaykrell** [comment]
Let's split up the F# tests here, and then fine tune the enable/disable, and then drill into the disabled. The failures are gsharedvt related, at least, but that doesn't make sense. The types should be statically discovered and instantiated and gsharedvt not used. Investigate more..

**Comment by jaykrell** [comment]
This is working well as intended -- splitting up the tests and then most pass, a few fail, some failures are understood (gsharedvt), some merit more investigation (x86). Still waiting on the pass/fail lists for arm..

File: mono/tests/Makefile.am [hunk 1]
```diff
@@ -912,7 +1009,106 @@ PLATFORM_DISABLED_TESTS += tailcall/3.exe
 PLATFORM_DISABLED_TESTS += tailcall/4.exe
 PLATFORM_DISABLED_TESTS += tailcall/8273.exe
 PLATFORM_DISABLED_TESTS += tailcall/fsharp-deeptail.exe
+PLATFORM_DISABLED_TESTS += tailcall/fsharp-deeptail-Seq.filter-length1.exe
+PLATFORM_DISABLED_TESTS += tailcall/fsharp-deeptail-Seq.filter-length2.exe
```
Comment (vargaz) [inline mono/tests/Makefile.am]: Could this be written in fewer lines ? The other places too.

**Comment by vargaz** [comment]
Also, would it be possible to move the decomposed tests into a subdir ? There are like a 100 files.

File: mono/tests/Makefile.am (hunk 1 above)
Comment (jaykrell) [reply mono/tests/Makefile.am]: It is this way for easy add/delete/merge. Adding a test is adding a line, removing a test is removing a line. No backslashes or word wrap to contend with.

**Comment by jaykrell** [comment]
Separate dir -- yes. I prefer that but much earlier @luhenry didn't seem to (he didn't really want the coreclr stuff as it is, but it was easier this way).
Tailcall/ or tailcall/x and tailcall/y or don't care as long as at least something/?
Probably just tailcall/fsharp-deeptail and tailcall/interface-conservestack or somesuch.

The "deeptail" and "conservestack" mean, like, vs. just at least running correctly.
deeptail means stack overflow for failure, though I go back and forth on recursion depth.
Some tests can detect stack consumption, some cannot. Perhaps not the best naming.
There were regressions of correctness at some point, which is why some tests run "just for correctness" w/o caring about stack use.

File: mono/tests/Makefile.am (hunk 1 above)
Comment (jaykrell) [reply mono/tests/Makefile.am]: The number of lines here should go down as well. I think I have commented out disable all for ARM, and disable all for PowerPC. Those should decrease. Oh and disable all for interpreter, granted. The *.il existing style is one per line, but with merge conflicts at the end due to lack of trailing slash. I believe ease of edit/merge is valuable, but it does conflict with reading quickly, glossing over stuff.

Maybe we can if-in the .il lines? That'd be a net reduction for certain.
We also maybe use include to shrink the file. But I'm not certain. At least the disables should reduce.

**Comment by jaykrell** [comment]
I've moved everything into subdirs tailcall/fsharp-deeptail/ and tailcall/interface-conservestack.
Reset all of the disabled lists, except what failed today. So that'll be smaller and hopefully not fully grow back. Will look at the x86 breaks.
Waiting for initial arm/arm64 lists from CI.
Will CI PowerPC also on next round.

**Comment by jaykrell** [comment]
x86 failures:
One is because called function does take more parameters than caller, which we don't support.

Some or all of the rest are related to frame alignment. The test checks for zero stack used, but 8 bytes were being used. I adjusted the tests -- take address of parameter instead of local, that helps some, maybe all, we'll see. The test also prints stack used in error message now.

The FullAOT failures are all believed to be due to gsharedvt.

Still waiting on ARM. :(

**Comment by jaykrell** [comment]
I also centralized the debugprinting.

**Comment by jaykrell** [comment]
Still waiting for ARM. Probably will have to redisable a very small number of tests due to extra_arg.

File: mono/tests/split-tailcall-interface-conservestack.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (luhenry) [inline mono/tests/split-tailcall-interface-conservestack.cpp]: Remove this file.

File: mono/tests/tailcall/split-fsharp.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/*
```
Comment (luhenry) [inline mono/tests/tailcall/split-fsharp.cpp]: Remove this file.

File: mono/tests/split-tailcall-interface-conservestack.cpp (hunk 2 above)
Comment (jaykrell) [reply mono/tests/split-tailcall-interface-conservestack.cpp]: This file is critical to maintenance of the tests. I have used it repeatedly as I edit the tests and it saved me a ton of time.

File: mono/tests/tailcall/split-fsharp.cpp (hunk 3 above)
Comment (jaykrell) [reply mono/tests/tailcall/split-fsharp.cpp]: This file is critical to maintenance of the tests. I have used it repeatedly as I edit the tests and it saved me a ton of time. I tried writing in Perl but gave up.

**Comment by jaykrell** [comment]
x86 passing with the test tweaks.
Overnight not long enough to run arm ci.

**Comment by jaykrell** [comment]
A few of the tests had the same names, so the last duplicate overwrote the others.
That is fixed locally, but it'd be good to get ARM results before resetting CI for another 18+ hours.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Looking good now. Review?

File: mono/tests/tailcall/split-fsharp.cpp (hunk 3 above)
Comment (vargaz) [reply mono/tests/tailcall/split-fsharp.cpp]: Its imho we already have test generating files like tests/gen-runtime-invoke.cs.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
I agree commiting derived files is generally bad. I’ll have to work the makefile & ensure the C++ works on centos6, or skip the tests there. And make sure it works out-of-tree. 

Some of the splitting is maybe redundant now, like all the interface tests maybe fail on arm but pass on the rest.

**Comment by jaykrell** [comment]
Shouldn’t be too difficult, I can try.

**Comment by luhenry** [comment]
Write the c++ in c#, that’s available everywhere we need it to be :wink:

**Comment by jaykrell** [comment]
Makefile not working yet.

**Comment by jaykrell** [comment]
@monojenkins Linux  WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT+LLVM

**Comment by jaykrell** [comment]
@monojenkins build OS X x64 iOS SDK

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7 Interpreter

File: mono/tests/tailcall/split-fsharp.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		suffix.push_back (line);
+
+	for (tests_t::const_iterator t = tests.begin(); t != tests.end(); ++t)
+	{
+		if (t->name.length() == 0)
+			continue;
+		//printf("%s\n", t->name.c_str());
+		ofstream output(("tailcall/fsharp-deeptail/" + t->name + ".il").c_str());
+		for (strings_t::const_iterator a = prefix.begin(); a != prefix.end(); ++a)
+			output << a->c_str () << endl;
```
Comment (jaykrell) [inline mono/tests/tailcall/split-fsharp.cpp]: c_str() not likely needed here

**Comment by lewurm** [comment]
Sorry for the strong language, but why the heck do we have C++ files now?!  This introduces another dependency that we don't want (we sort of need it for LLVM, but this is optional). This was sneakily introduced after the initial PR approval. For the future, if you do larger changes on your PR _after_ it was approved, please dismiss the review and ask for review again.

Can you please rewrite those generators in C# as suggested in https://github.com/mono/mono/pull/9316#issuecomment-402266442 ?

**Comment by vargaz** [comment]
The file is not part of the build, its just used to generate the original versions of tests, so its not a big problem.

**Comment by jaykrell** [comment]
It is used now.  @luhenry kinda suggested that -- if keeping file, then use it. And then the rewrite. I took only the first suggestion.

**Pull request merged**